File: lib/py/graph_basic.py
from collections import deque

# DFS (재귀)
def dfs(cur, adj, visited):
    visited[cur] = True
    print(cur, end=' ')
    for nxt in adj[cur]:
        if not visited[nxt]:
            dfs(nxt, adj, visited)

# BFS (큐)
def bfs(start, adj, visited):
    q = deque([start])
    visited[start] = True
    
    while q:
        cur = q.popleft()
        print(cur, end=' ')
        
        for nxt in adj[cur]:
            if not visited[nxt]:
                visited[nxt] = True
                q.append(nxt)

# 위상 정렬
def topological_sort(n, adj, indegree):
    q = deque([i for i in range(1, n + 1) if indegree[i] == 0])
    result = []
    
    while q:
        cur = q.popleft()
        result.append(cur)
        
        for nxt in adj[cur]:
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                q.append(nxt)
                
    if len(result) != n:
        return [] # 사이클 존재
    return result

File: lib/cpp/graph_basic.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>


using namespace std;

const int MAX = 100005;
vector<int> adj[MAX];
bool visited[MAX];

// DFS (재귀)
void dfs(int cur) {
  visited[cur] = true;
  cout << cur << " ";
  for (int next : adj[cur]) {
    if (!visited[next]) {
      dfs(next);
    }
  }
}

// BFS (큐)
void bfs(int start) {
  queue<int> q;
  q.push(start);
  visited[start] = true;

  while (!q.empty()) {
    int cur = q.front();
    q.pop();
    cout << cur << " ";

    for (int next : adj[cur]) {
      if (!visited[next]) {
        visited[next] = true;
        q.push(next);
      }
    }
  }
}

// 위상 정렬 (Kahn's Algorithm)
vector<int> topologicalSort(int n, vector<int> &indegree) {
  queue<int> q;
  vector<int> result;

  for (int i = 1; i <= n; i++) {
    if (indegree[i] == 0)
      q.push(i);
  }

  while (!q.empty()) {
    int cur = q.front();
    q.pop();
    result.push_back(cur);

    for (int next : adj[cur]) {
      indegree[next]--;
      if (indegree[next] == 0)
        q.push(next);
    }
  }

  if (result.size() != n)
    return {}; // 사이클 존재
  return result;
}
